pwiz_tools/Skyline/Model/AuditLog/AuditLogEntry.cs

```diff
@@ -300,24 +328,36 @@ public static bool ReadFromXmlTextReader(XmlTextReader reader, out string logged
                 }
             }
 
-            string rootHash = null;
+            string rootHashString = null;
             if (reader.IsStartElement(EL.root_hash.ToString()))
-                rootHash = reader.ReadElementString(EL.root_hash.ToString());
+                rootHashString = reader.ReadElementString(EL.root_hash.ToString());
 
             result = reader.DeserializeElement<AuditLogList>();
             result.FormatVersion = docFormat;
             result.RootHash = new AuditLogHash()
                 .ChangeActualHash(result.CalculateRootHash())
-                .ChangeSkylHash(Hash.FromBase64(rootHash));
+                .ChangeSkylHash(Hash.FromBase64(rootHashString));
 
-            // yell if audit log entry hash does not match.
-            // If the docFormat is null, this is an old audit log and there won't be any entry hashes.
-            // We can't just always ignore non-existent hashes, otherwise people could just delete the hash elements
+            // Root and entry-level hashes validation
+            // We can't ignore non-existent hashes, otherwise people could just delete the hash elements
             // and get Skyline to successfully load the audit log
-            var modifiedEntries =
-                result.AuditLogEntries.Enumerate().Where(entry => !entry.Hash.SkylAndActualHashesEqual()).ToArray();
 
-            if (docFormat != null && (!result.RootHash.SkylAndActualHashesEqual() || modifiedEntries.Length > 0))
+            // recalculate the actual hash for the current document format
+            var getHashes = new Func<AuditLogEntry, KeyValuePair<AuditLogHash, AuditLogEntry>>(e =>new KeyValuePair<AuditLogHash, AuditLogEntry>(e.Hash, e));
+            if (!DocumentFormat.CURRENT.Equals(result.FormatVersion))
+                getHashes = new Func<AuditLogEntry, KeyValuePair<AuditLogHash, AuditLogEntry>>(
+                    e => new KeyValuePair<AuditLogHash, AuditLogEntry>(new AuditLogHash(e, e.Hash.SkylHash, docFormat), e));
+
+            var hashes = result.AuditLogEntries.Enumerate().Select(getHashes);
```

Comment:
Can you change this to a "foreach" with an if statement inside of it?
I think  that would be much easier to undertand compared to having "getHashes" being a Func<AuditLogEntry...>.

Reply:
Yes, I rewrote this piece, it was a bit clumsy.